File: central_flutter/lib/ble/ble_transport.dart
import 'dart:async';
import 'dart:collection';
import 'dart:io' show Platform;

import 'package:flutter/foundation.dart';
import 'package:flutter_blue_plus/flutter_blue_plus.dart';

const String serviceUuid = '12340001-0000-1000-8000-00805f9b34fb';
const String charUuid = '12340002-0000-1000-8000-00805f9b34fb';

class ScannedDevice {
  final BluetoothDevice device;
  final String? name;
  final String address;

  ScannedDevice({required this.device, this.name, required this.address});
}

class BleTransport {
  BluetoothDevice? _device;
  BluetoothCharacteristic? _char;
  StreamSubscription<List<int>>? _notifySub;
  final _notifyQueue = Queue<Uint8List>();
  Completer<Uint8List>? _notifyWaiter;
  int _mtu = 23;

  int get mtu => _mtu;

  Future<List<ScannedDevice>> scan({
    Duration timeout = const Duration(seconds: 5),
    String? filterServiceUuid,
  }) async {
    final allDevices = <String, ScanResult>{};

    // Use unfiltered scan and filter manually — Android's ScanFilter
    // does not reliably match 128-bit service UUIDs on all devices.
    final sub = FlutterBluePlus.onScanResults.listen((results) {
      for (final r in results) {
        allDevices[r.device.remoteId.str] = r;
      }
    });

    await FlutterBluePlus.startScan(timeout: timeout);
    await FlutterBluePlus.isScanning.firstWhere((s) => !s);
    sub.cancel();

    debugPrint('[BLE] Scan complete: ${allDevices.length} device(s) found');
    for (final r in allDevices.values) {
      final name = r.device.platformName.isNotEmpty
          ? r.device.platformName
          : r.advertisementData.advName;
      debugPrint(
          '[BLE]   ${r.device.remoteId.str} "$name" services=${r.advertisementData.serviceUuids.map((u) => u.str).toList()}');
    }

    final targetGuid = Guid(filterServiceUuid ?? serviceUuid);
    final devices = <ScannedDevice>[];
    for (final r in allDevices.values) {
      final hasService =
          r.advertisementData.serviceUuids.any((u) => u == targetGuid);
      if (!hasService) continue;
      devices.add(ScannedDevice(
        device: r.device,
        name: r.device.platformName.isNotEmpty
            ? r.device.platformName
            : r.advertisementData.advName.isNotEmpty
                ? r.advertisementData.advName
                : null,
        address: r.device.remoteId.str,
      ));
    }
    return devices;
  }

  Future<void> connect(ScannedDevice scannedDevice) async {
    _device = scannedDevice.device;
    await _device!.connect(autoConnect: false, mtu: null);

    // Request MTU (Android only — iOS negotiates MTU automatically)
    if (Platform.isAndroid) {
      _mtu = await _device!.requestMtu(247);
    } else {
      _mtu = _device!.mtuNow;
    }

    // Discover services
    final services = await _device!.discoverServices();
    BluetoothCharacteristic? targetChar;
    for (final s in services) {
      if (s.uuid == Guid(serviceUuid)) {
        for (final c in s.characteristics) {
          if (c.uuid == Guid(charUuid)) {
            targetChar = c;
            break;
          }
        }
      }
    }
    if (targetChar == null) {
      throw StateError('blerpc characteristic not found');
    }
    _char = targetChar;

    // Enable notifications — use a queue so that events arriving between
    // consecutive readNotify() calls are buffered instead of lost.
    _notifyQueue.clear();
    _notifyWaiter = null;
    await _char!.setNotifyValue(true);
    _notifySub = _char!.onValueReceived.listen((value) {
      final data = Uint8List.fromList(value);
      if (_notifyWaiter != null && !_notifyWaiter!.isCompleted) {
        _notifyWaiter!.complete(data);
        _notifyWaiter = null;
      } else {
        _notifyQueue.add(data);
      }
    });
  }

  Future<void> write(Uint8List data) async {
    if (_char == null) throw StateError('Not connected');
    await _char!.write(data, withoutResponse: true);
  }

  Future<Uint8List> readNotify({Duration? timeout}) async {
    if (_char == null) throw StateError('Not connected');
    if (_notifyQueue.isNotEmpty) {
      return _notifyQueue.removeFirst();
    }
    final waiter = Completer<Uint8List>();
    _notifyWaiter = waiter;
    timeout ??= const Duration(seconds: 2);
    try {
      return await waiter.future.timeout(timeout);
    } on TimeoutException {
      _notifyWaiter = null;
      rethrow;
    }
  }

  /// Drain any pending notifications.
  Future<void> drainNotifications() async {
    _notifyQueue.clear();
    // Also wait briefly for any in-flight notifications
    try {
      while (true) {
        await readNotify(timeout: const Duration(milliseconds: 100));
      }
    } on TimeoutException {
      // Done draining
    }
  }

  void disconnect() {
    _notifySub?.cancel();
    _notifySub = null;
    _notifyQueue.clear();
    _notifyWaiter = null;
    _device?.disconnect();
    _device = null;
    _char = null;
  }
}

File: central_rn/src/ble/BleTransport.ts
import { BleManager, Device, Characteristic, Subscription } from 'react-native-ble-plx';
import { Buffer } from 'buffer';

export const SERVICE_UUID = '12340001-0000-1000-8000-00805f9b34fb';
export const CHAR_UUID = '12340002-0000-1000-8000-00805f9b34fb';

export interface ScannedDevice {
  device: Device;
  name: string | null;
  address: string;
}

export class BleTransport {
  private _manager: BleManager;
  private _device: Device | null = null;
  private _char: Characteristic | null = null;
  private _notifySub: Subscription | null = null;
  private _notifyQueue: Uint8Array[] = [];
  private _notifyWaiter: {
    resolve: (value: Uint8Array) => void;
    reject: (reason: Error) => void;
  } | null = null;
  private _mtu = 23;

  constructor() {
    this._manager = new BleManager();
  }

  get mtu(): number {
    return this._mtu;
  }

  async scan(timeout = 5000): Promise<ScannedDevice[]> {
    const allDevices = new Map<string, Device>();

    return new Promise<ScannedDevice[]>((resolve) => {
      this._manager.startDeviceScan(null, null, (error, device) => {
        if (error) {
          console.warn('[BLE] Scan error:', error.message);
          return;
        }
        if (device) {
          allDevices.set(device.id, device);
        }
      });

      setTimeout(() => {
        this._manager.stopDeviceScan();
        console.log(`[BLE] Scan complete: ${allDevices.size} device(s) found`);

        const devices: ScannedDevice[] = [];
        for (const device of allDevices.values()) {
          const serviceUUIDs = device.serviceUUIDs ?? [];
          const hasService = serviceUUIDs.some(
            (u) => u.toLowerCase() === SERVICE_UUID.toLowerCase(),
          );
          if (!hasService) continue;
          devices.push({
            device,
            name: device.name ?? device.localName ?? null,
            address: device.id,
          });
        }
        resolve(devices);
      }, timeout);
    });
  }

  async connect(scannedDevice: ScannedDevice): Promise<void> {
    this._device = await scannedDevice.device.connect({ requestMTU: 247 });

    // Get negotiated MTU
    this._mtu = this._device.mtu ?? 23;

    // Discover services and characteristics
    await this._device.discoverAllServicesAndCharacteristics();
    const services = await this._device.services();
    let targetChar: Characteristic | null = null;

    for (const service of services) {
      if (service.uuid.toLowerCase() === SERVICE_UUID.toLowerCase()) {
        const chars = await service.characteristics();
        for (const c of chars) {
          if (c.uuid.toLowerCase() === CHAR_UUID.toLowerCase()) {
            targetChar = c;
            break;
          }
        }
      }
    }

    if (!targetChar) {
      throw new Error('blerpc characteristic not found');
    }
    this._char = targetChar;

    // Enable notifications with queue
    this._notifyQueue = [];
    this._notifyWaiter = null;
    this._notifySub = this._char.monitor((error, characteristic) => {
      if (error) {
        console.warn('[BLE] Notify error:', error.message);
        return;
      }
      if (!characteristic?.value) return;

      const data = new Uint8Array(Buffer.from(characteristic.value, 'base64'));
      if (this._notifyWaiter) {
        const waiter = this._notifyWaiter;
        this._notifyWaiter = null;
        waiter.resolve(data);
      } else {
        this._notifyQueue.push(data);
      }
    });
  }

  async write(data: Uint8Array): Promise<void> {
    if (!this._char) throw new Error('Not connected');
    const base64 = Buffer.from(data).toString('base64');
    await this._char.writeWithoutResponse(base64);
  }

  async readNotify(timeout = 2000): Promise<Uint8Array> {
    if (!this._char) throw new Error('Not connected');
    if (this._notifyQueue.length > 0) {
      return this._notifyQueue.shift()!;
    }
    return new Promise<Uint8Array>((resolve, reject) => {
      const timer = setTimeout(() => {
        this._notifyWaiter = null;
        reject(new Error('Timeout waiting for notification'));
      }, timeout);

      this._notifyWaiter = {
        resolve: (value) => {
          clearTimeout(timer);
          resolve(value);
        },
        reject: (reason) => {
          clearTimeout(timer);
          reject(reason);
        },
      };
    });
  }

  async drainNotifications(): Promise<void> {
    this._notifyQueue = [];
    try {
      for (;;) {
        await this.readNotify(100);
      }
    } catch {
      // Done draining
    }
  }

  disconnect(): void {
    this._notifySub?.remove();
    this._notifySub = null;
    this._notifyQueue = [];
    this._notifyWaiter = null;
    if (this._device) {
      this._device.cancelConnection().catch(() => {});
      this._device = null;
    }
    this._char = null;
  }
}
